library/src/main/java/com/datatorrent/lib/iteration/DelayByNOperator.java

```diff
@@ -0,0 +1,225 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.datatorrent.lib.iteration;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import javax.validation.constraints.Min;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.Operator;
+import com.datatorrent.lib.util.WindowDataManager;
+import com.datatorrent.netlet.util.DTThrowable;
+import com.datatorrent.stram.engine.WindowGenerator;
+
+/**
+ * This operator is used in iteration and can do n window delay.
+ * Required delay needs to be set in the constructor, default delay is 1.
+  *
+ * @displayName nDelayOperator
+ * @category iteration
+
+ */
+public class DelayByNOperator<T> implements Operator.DelayOperator, Operator.CheckpointListener
+{
+  private WindowDataManager windowDataManager = new WindowDataManager.FSWindowDataManager();
+  @Min(1)
+  private int delay = 1;
+  private long currentWindowId;
+  private transient int operatorContextId;
+  private transient ArrayList<T> windowData;
+  private transient Context.OperatorContext context;
+  private long previouslyCommittedWindow;
+  private boolean previouslyCommittedWindowSet = false;
+
+  public transient DefaultInputPort<T> input = new DefaultInputPort<T>() {
+    @Override
+    public void process(T t)
+    {
+      processTuple(t);
+    }
+  };
+
+  public transient DefaultOutputPort<T> output = new DefaultOutputPort();
+
+  public DelayByNOperator()
+  {
+    init();
+  }
+
+  /*
+  * @param delay set the delay in number of windows for the tuples.
+   */
+  public DelayByNOperator(int delay)
+  {
+    if ( delay < 1 ) {
```

Comment:
When the logic is done, please correct the code style so it can be merged after https://github.com/apache/incubator-apex-core/pull/185 is merged.

- no space after open parenthesis and before close parenthesis.
- classes (including unit test classes) must be capitalized.
- (someBoolean == false) should be changed to (!someBoolean)
- all open braces that denotes a definition must be in its own line